commit e21546a2a3953a7d4b9d5c9b1cf12184ffceab96
Author: Johannes Berg <johannes@sipsolutions.net>
Date:   Thu Aug 6 20:41:32 2009 +0200

    mac80211: stay authenticated after disassoc
    
    After being disassociated by the AP, mac80211 currently
    reports this to cfg80211, and then goes to delete the
    association. That's fine, but cfg80211 assumes that it's
    still authenticated, however, mac80211 throws away all
    state.
    
    This fixes mac80211 to keep track of the authentication
    in that case so that cfg80211 can request a deauth or
    new association properly.
    
    Signed-off-by: Johannes Berg <johannes@sipsolutions.net>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index 630a438180fd..989591787aee 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -284,6 +284,7 @@ struct ieee80211_if_managed {
 
 	struct mutex mtx;
 	struct ieee80211_bss *associated;
+	struct ieee80211_mgd_work *old_associate_work;
 	struct list_head work_list;
 
 	u8 bssid[ETH_ALEN];